fairlearn/metrics/_metric_frame.py

```diff
@@ -117,24 +129,156 @@ def __init__(self,
+            assert y_true == [] and y_pred == [] and sensitive_features == [], _NON_EMPTY_INIT_ERR
+            self._y_true = []
+            self._y_pred = []
+            self._overall = None
+            self._by_group = None
+
+    def add_data(self,
+                 y_true,
+                 y_pred, *,
+                 sensitive_features,
+                 control_features: Optional = None,
+                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None
+                 ):
+        """Add data to the MetricFrame.
+
+        Parameters
+        ----------
+        y_true : List, pandas.Series, numpy.ndarray, pandas.DataFrame
+        The ground-truth labels (for classification) or target values (for regression).
+
+        y_pred : List, pandas.Series, numpy.ndarray, pandas.DataFrame
+                The predictions.
+
+        sensitive_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray,
+         pandas.DataFrame
+            The sensitive features which should be used to create the subgroups.
+            At least one sensitive feature must be provided.
+            All names (whether on pandas objects or dictionary keys) must be strings.
+            We also forbid DataFrames with column names of ``None``.
+            For cases where no names are provided we generate names ``sensitive_feature_[n]``.
+
+        control_features : List, pandas.Series, dict of 1d arrays, numpy.ndarray, DataFrame
+            Control features are similar to sensitive features, in that they
+            divide the input data into subgroups.
+            Unlike the sensitive features, aggregations are not performed
+            across the control features - for example, the ``overall`` property
+            will have one value for each subgroup in the control feature(s),
+            rather than a single value for the entire data set.
+            Control features can be specified similarly to the sensitive features.
+            However, their default names (if none can be identified in the
+            input values) are of the format ``control_feature_[n]``.
+
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+
+        sample_params : dict
+            Parameters for the metric function(s). If there is only one metric function,
+            then this is a dictionary of strings and array-like objects, which are split
+            alongside the ``y_true`` and ``y_pred`` arrays, and passed to the metric function.
+            If there are multiple metric functions (passed as a dictionary), then this is
+            a nested dictionary, with the first set of string keys identifying the
+            metric function name, with the values being the string-to-array-like dictionaries.
+        """
+        if not self._streaming:
+            raise Exception("This MetricFrame does not support adding data.")
```

Comment:
Again, not vital quite yet, but could we raise a more specific exception? Although I'm not sure that `ValueError` is appropriate here.

Reply:
done

Reply:
@adrinjalali  would have better thoughts than I do on that.

Reply:
Isn't this a `NotImplementedError`?